contrib/endpoints/src/api_manager/config_test.cc

```diff
@@ -870,6 +870,80 @@ TEST(Config, TestCorsDisabled) {
   ASSERT_EQ(nullptr, method1);
 }
 
+TEST(Config, TestInvalidMetricRules) {
+  MockApiManagerEnvironmentWithLog env;
+  // There is no http.rule or api.method to match the selector.
+  static const char config_text[] = R"(
+name: "Service.Name"
+quota {
+   metric_rules {
+      selector: "GetShelves"
+      metric_costs {
+         key: "test.googleapis.com/operation/read_book"
+         value: 100
+      }
+   }
+}
+)";
+
+  std::unique_ptr<Config> config = Config::Create(&env, config_text, "");
+  EXPECT_EQ(nullptr, config);
+}
+
+TEST(Config, TestMetricRules) {
+  MockApiManagerEnvironmentWithLog env;
+  static const char config_text[] = R"(
+name: "Service.Name"
+http {
+  rules {
+   selector: "DeleteShelf"
+   delete: "/shelves"
+  }
+  rules {
+   selector: "GetShelves"
+   get: "/shelves"
+  }
+}
+quota {
+   metric_rules {
+      selector: "GetShelves"
+      metric_costs {
+         key: "test.googleapis.com/operation/get_shelves"
+         value: 100
+      }
+   }
+   metric_rules {
+      selector: "DeleteShelf"
+      metric_costs {
+         key: "test.googleapis.com/operation/delete_shelves"
+         value: 200
+      }
+   }
+}
+)";
+
+  std::unique_ptr<Config> config = Config::Create(&env, config_text, "");
+  ASSERT_TRUE(config);
+
+  auto method1 = config->GetMethodInfo("GET", "/shelves");
```

Comment:
To keep the code consistent with Envoy, I suggest avoiding over usage of auto method. In the Envoy codebase, we use explicit types most of the time